core/src/main/scala/kafka/idempotence/ProducerIdMapping.scala

```diff
@@ -0,0 +1,333 @@
+/**
+  * Licensed to the Apache Software Foundation (ASF) under one or more
+  * contributor license agreements.  See the NOTICE file distributed with
+  * this work for additional information regarding copyright ownership.
+  * The ASF licenses this file to You under the Apache License, Version 2.0
+  * (the "License"); you may not use this file except in compliance with
+  * the License.  You may obtain a copy of the License at
+  *
+  * http://www.apache.org/licenses/LICENSE-2.0
+  *
+  * Unless required by applicable law or agreed to in writing, software
+  * distributed under the License is distributed on an "AS IS" BASIS,
+  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  * See the License for the specific language governing permissions and
+  * limitations under the License.
+  */
+
+package kafka.idempotence
+
+import java.io._
+import java.nio.ByteBuffer
+import java.nio.file.Files
+
+import kafka.common.{KafkaException, TopicAndPartition}
+import kafka.log.LogConfig
+import kafka.utils.{Logging, nonthreadsafe}
+import org.apache.kafka.common.errors.{InvalidSequenceNumberException, ProducerFencedException}
+import org.apache.kafka.common.protocol.types._
+import org.apache.kafka.common.utils.Utils
+
+import scala.collection.mutable
+
+private[idempotence] case class PidEntry(seq: Int, epoch: Short, offset: Long)
+private[idempotence] class CorruptSnapshotException(msg: String) extends KafkaException(msg)
+
+object ProducerIdMapping {
+  private val DirnamePrefix = "pid-mapping-"
+  private val FilenameSuffix = "snapshot"
+  private val FilenamePattern = s"^\\d{1,}.$FilenameSuffix".r
+  private val PidSnapshotVersion: Short = 1
+
+  private val VersionField = "version"
+  private val CrcField = "crc"
+  private val PidField = "pid"
+  private val SequenceField = "sequence"
+  private val EpochField = "epoch"
+  private val OffsetField = "offset"
+  private val PidEntriesField = "pid_entries"
+
+  private val VersionOffset = 0
+  private val CrcOffset = VersionOffset + 2
+  private val PidEntriesOffset = CrcOffset + 4
+
+  val PidSnapshotEntrySchema = new Schema(
+    new Field(PidField, Type.INT64, "The producer ID"),
+    new Field(EpochField, Type.INT16, "Current epoch of the producer"),
+    new Field(SequenceField, Type.INT32, "Last written sequence of the producer"),
```

Comment:
Discussed offline, but for posterity: Using 64 bits would probably be ideal. Handling sequence number wraparound is tricky because of the need to detect duplicates (e.g. if we wrap around to sequence number 0, then sequence numbers "close" to `Integer.MAX_VALUE` should probably be seen as duplicates). 

Another option is to bump the epoch whenever we reach the limit for a partition. This sounds nice in theory, but has some bad implications for transactions. In particular, we may find the need to bump the epoch while a transaction is in progress, which is impossible in the current design.

We probably need to think about this a bit more and decide if saving the 4 bytes is worth it.